provider_shopper/test/catalog_widget_test.dart

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019 The Flutter team. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:provider_shopper/main.dart';
+import 'package:flutter/material.dart';
+import 'package:provider_shopper/models/catalog.dart';
+
+void main() {
+  TestWidgetsFlutterBinding.ensureInitialized();
+  testWidgets('Catalog page Widget test', (tester) async {
+    await tester.pumpWidget(MyApp());
+
+    // Navigating through login page.
+    await tester.tap(find.text('ENTER'));
+    await tester.pumpAndSettle();
+
+// Now we are at the Catalog page.
+    expect(find.byType(CustomScrollView), findsOneWidget);
+
+    expect(find.text('Catalog'), findsOneWidget);
+
+    expect(find.byIcon(Icons.shopping_cart), findsOneWidget);
+
+    expect(
+        find.ancestor(
+            of: find.text('Catalog'), matching: find.byType(SliverAppBar)),
+        findsOneWidget);
+
+    await tester.tap(find.byIcon(Icons.shopping_cart));
+    await tester.pumpAndSettle();
+    expect(find.text('Catalog'), findsNothing);
+
+    await tester.pageBack();
+    await tester.pumpAndSettle();
+
+    expect(find.byType(SliverToBoxAdapter), findsOneWidget);
+    expect(find.byType(SliverList), findsOneWidget);
+
+    final catalogListItems = CatalogModel.getCatalogItems();
+
+    for (String item in catalogListItems.sublist(0, 9)) {
+      expect(find.text(item), findsOneWidget);
+      expect(
+          find.ancestor(of: find.text(item), matching: find.byType(Expanded)),
+          findsOneWidget);
+    }
+
+    await tester.drag(find.byType(CustomScrollView), const Offset(0.0, -380.0));
+    await tester.pumpAndSettle();
+
+    for (String item in catalogListItems.sublist(9, catalogListItems.length)) {
+      expect(find.text(item), findsOneWidget);
+      expect(
+          find.ancestor(of: find.text(item), matching: find.byType(Expanded)),
+          findsOneWidget);
+    }
+
+    await tester.drag(find.byType(CustomScrollView), const Offset(0.0, 0.0));
+    await tester.pumpAndSettle();
+
+    // Should find 11 buttons at the current/base viewport height.
+    expect(find.text('ADD'), findsNWidgets(11));
+
+    await tester.tap(find.widgetWithText(FlatButton, 'ADD').first);
+    await tester.pumpAndSettle();
+
+    // expect(find.byIcon(Icons.check), findsOneWidget);
```

Comment:
@RedBrogdon Also wanted to confirm something with you. This statement should successfully pass right given the previous actions. But surprisingly this was failing :thinking: Could you give me any insight?

Reply:
Hey @jddeep, I guess the issue is with the scrolling which you are doing. I have a few doubts when performing drag you are giving the `Offest(0, -380)`. How did you calculate that?

If I give the `Offset(0, -350)`, and instead of finding exactly 11 widgets, if I look for N widgets, the test is passing. The funny thing is if I give `Offset(0, -360)` the test will fail. And if I tap on the last instead of first `FlatButton` with ADD the test is passing.

Here's a proposed solution to avoid that, if you want to check if the scrolling is working properly you can do like this.

``` dart

// Perform Scroll Up
    await tester.drag(find.byType(CustomScrollView), const Offset(0.0, -360.0));
    await tester.pumpAndSettle();
    
    // Once scroll up is performed, we should not be able to find Catalog
    expect(find.text('Catalog'),findsNothing);
    
    // Perform Scroll down
    await tester.drag(find.byType(CustomScrollView), const Offset(0.0, 360.0));
    await tester.pumpAndSettle();
    
    // Once scroll down is performed, we should be able to find Catalog
    expect(find.text('Catalog'),findsOneWidget);
    
    // Should find multiple widget
    expect(find.text('ADD'), findsWidgets);
    await tester.tap(find.text('ADD').first);
    await tester.pumpAndSettle();

    expect(find.byIcon(Icons.check), findsOneWidget);
```

Reply:
@AyushBherwani1998 actually, I got that offset value according to the viewport height and also by dry running the test and evaluating my desired/expected results. Everything else is working as expected cause I also wanted to test the list items(the item names) and how they are laid out. Only testing the scrolling was not my actual and sole aim.
After tapping the `first` - `ADD` button, there should be `one` - `check` icon by means of simple logic. But that particular expectation was not being satisfied. That's what got me thinking :thinking:

Reply:
Exactly, the logic should work. I did try the same locally but not able to exactly identify the issue.

Reply:
Yeah cc: @RedBrogdon